File: lib/libsokoengine/src/board/sokoban_plus.cpp
#include "sokoban_plus.hpp"
#include "text_utils.hpp"

#include <map>
#include <set>
#include <vector>

#include <boost/algorithm/string.hpp>
#include <boost/lexical_cast.hpp>

using namespace std;
using namespace boost;

namespace sokoengine {

SokobanPlusDataError::SokobanPlusDataError(const string& mess):
  invalid_argument(mess)
{}

SokobanPlusDataError::~SokobanPlusDataError() = default;

KeyError::KeyError(const string& mess):
  invalid_argument(mess)
{}

KeyError::~KeyError() = default;

class LIBSOKOENGINE_LOCAL SokobanPlus::PIMPL {
public:
  typedef map<piece_id_t, piece_id_t> IdsMap;

  IdsMap m_box_plus_ids;
  IdsMap m_goal_plus_ids;
  string m_boxorder;
  string m_goalorder;
  size_t m_pieces_count = 0;
  bool m_enabled = false;
  bool m_validated = false;
  vector<string> m_errors;

  enum {
    LEGACY_DEFAULT_PLUS_ID = 99
  };

  piece_id_t get_plus_id(piece_id_t for_id, const IdsMap& from_where) const {
    if(m_enabled) {
      return from_where.at(for_id);
    } else {
      return DEFAULT_PLUS_ID;
    }
  }

  string rstrip_default_plus_ids(const string& str) const {
    if (m_pieces_count < LEGACY_DEFAULT_PLUS_ID) {
      return trim_right_copy_if(
        str,
        boost::is_any_of(lexical_cast<string>(LEGACY_DEFAULT_PLUS_ID)) ||
        boost::is_any_of(lexical_cast<string>(SokobanPlus::DEFAULT_PLUS_ID)) ||
        boost::is_space()
      );
    } else {
      return trim_right_copy_if(
        str,
        boost::is_any_of(
          lexical_cast<string>(SokobanPlus::DEFAULT_PLUS_ID)
        ) || boost::is_space()
      );
    }
  }

  IdsMap parse_and_clean_ids_string(const string& ids_str) const {
    vector<string> trimmed;
    string tmp = rstrip_default_plus_ids(ids_str);
    boost::algorithm::split(
      trimmed, tmp, is_space(),
      boost::token_compress_on
    );

    vector<piece_id_t> cleaned;
    for(const string& str_id : trimmed) {
      if (!TextUtils::is_blank(str_id)) {
        piece_id_t converted_id;
        try {
          converted_id = lexical_cast<piece_id_t>(str_id);
        }
        catch(const bad_lexical_cast&) {
          throw SokobanPlusDataError(
            "Can't parse Sokoban+ string! Illegal characters found. Only digits and spaces allowed."
          );
        }

        if (converted_id == LEGACY_DEFAULT_PLUS_ID &&
            m_pieces_count < LEGACY_DEFAULT_PLUS_ID) {
          converted_id = SokobanPlus::DEFAULT_PLUS_ID;
        }

        cleaned.push_back(converted_id);
      }
    }

    if (cleaned.size() < m_pieces_count) {
      cleaned.resize(m_pieces_count, SokobanPlus::DEFAULT_PLUS_ID);
    }

    IdsMap retv;
    piece_id_t index = 0;
    for(const auto& plus_id : cleaned) {
      retv[index + DEFAULT_PIECE_ID] = plus_id;
      index++;
    }

    return retv;
  }

  void validate_plus_ids(const IdsMap& ids) {
    for (auto id : ids) {
      if (!SokobanPlus::is_valid_plus_id(id.second)) {
        m_errors.push_back("Invalid Sokoban+ ID: " + std::to_string(id.second));
      }
    }
  }

  void validate_ids_counts() {
    if (!m_box_plus_ids.empty() && m_box_plus_ids.size() != m_pieces_count) {
      m_errors.push_back(
        string("Sokoban+ boxorder data doesn't contain same amount ") +
        "of IDs as there are pieces on board!. (pieces_count: " +
        std::to_string(m_pieces_count) + ")"
      );
    }

    if (!m_goal_plus_ids.empty() && m_goal_plus_ids.size() != m_pieces_count) {
      m_errors.push_back(
        string("Sokoban+ goalorder data doesn't contain same amount ") +
        "of IDs as there are pieces on board!. (pieces_count: " +
        std::to_string(m_pieces_count) + ")"
      );
    }
  }

  void validate_id_sets_equality() {
    set<piece_id_t> boxes, goals;
    for (auto id : m_box_plus_ids) {
      if (id.second != SokobanPlus::DEFAULT_PLUS_ID) {
        boxes.insert(id.second);
      }
    }
    for (auto id : m_goal_plus_ids) {
      if (id.second != SokobanPlus::DEFAULT_PLUS_ID) {
        goals.insert(id.second);
      }
    }

    if (boxes != goals) {
      m_errors.push_back(
        "Sokoban+ data doesn't define equal sets of IDs for boxes and goals"
      );
    }
  }
};

SokobanPlus::SokobanPlus() :
  m_impl(make_unique<PIMPL>())
{}

SokobanPlus::SokobanPlus(size_t pieces_count, const string& boxorder,
                        const string& goalorder) :
  SokobanPlus()
{
  m_impl->m_pieces_count = pieces_count;
  m_impl->m_boxorder = boxorder;
  m_impl->m_goalorder = goalorder;
}

SokobanPlus::SokobanPlus(const SokobanPlus& rv) :
  m_impl(make_unique<PIMPL>(*rv.m_impl))
{}

SokobanPlus& SokobanPlus::operator=(const SokobanPlus& rv) {
  if (this != &rv) {
      m_impl = make_unique<PIMPL>(*rv.m_impl);
  }
  return *this;
}

SokobanPlus::SokobanPlus(SokobanPlus &&) = default;

SokobanPlus& SokobanPlus::operator=(SokobanPlus &&) = default;

SokobanPlus::~SokobanPlus() = default;

bool SokobanPlus::operator== (const SokobanPlus& rv) const {
  return m_impl->m_pieces_count == rv.m_impl->m_pieces_count &&
         boxorder() == rv.boxorder() &&
         goalorder() == rv.goalorder();
}

bool SokobanPlus::operator!= (const SokobanPlus& rv) const {
  return !(*this == rv);
}

size_t SokobanPlus::pieces_count() const {
  return m_impl->m_pieces_count;
}

void SokobanPlus::set_pieces_count(size_t rv) {
  if (rv != m_impl->m_pieces_count) {
    disable();
    m_impl->m_validated = false;
    m_impl->m_pieces_count = rv;
  }
}

string SokobanPlus::boxorder() const {
  if (is_enabled() && is_valid()) {
    vector<string> tmp;
    for (const auto& id : m_impl->m_box_plus_ids) {
      tmp.push_back(std::to_string(id.second));
    }
    string tmp2 = boost::algorithm::join(tmp, " ");
    return m_impl->rstrip_default_plus_ids(tmp2);
  } else {
    return m_impl->m_boxorder;
  }
}

void SokobanPlus::set_boxorder(const string& rv) {
  if (rv != m_impl->m_boxorder) {
    disable();
    m_impl->m_validated = false;
    m_impl->m_boxorder = rv;
  }
}

string SokobanPlus::goalorder() const {
  if (is_enabled() and is_valid()) {
    vector<string> tmp;
    for (const auto& id : m_impl->m_goal_plus_ids) {
      tmp.push_back(std::to_string(id.second));
    }
    string tmp2 = boost::algorithm::join(tmp, " ");
    return m_impl->rstrip_default_plus_ids(tmp2);
  } else {
    return m_impl->m_goalorder;
  }
}

void SokobanPlus::set_goalorder(const string& rv) {
  if (rv != m_impl->m_goalorder) {
    disable();
    m_impl->m_validated = false;
    m_impl->m_goalorder = rv;
  }
}

bool SokobanPlus::is_valid() const {
  if (m_impl->m_validated == false) {
    m_impl->m_validated = true;
    m_impl->m_errors.clear();

    try {
      m_impl->m_box_plus_ids = m_impl->parse_and_clean_ids_string(m_impl->m_boxorder);
      m_impl->m_goal_plus_ids = m_impl->parse_and_clean_ids_string(m_impl->m_goalorder);
    }
    catch (const std::exception& exc) {
      m_impl->m_errors.push_back(exc.what());
    }

    m_impl->validate_plus_ids(m_impl->m_box_plus_ids);
    m_impl->validate_plus_ids(m_impl->m_goal_plus_ids);
    m_impl->validate_ids_counts();
    m_impl->validate_id_sets_equality();
  }

  return m_impl->m_errors.empty();
}

bool SokobanPlus::is_enabled() const {
  return m_impl->m_enabled;
}

void SokobanPlus::enable() {
  if (!is_valid()) {
    throw SokobanPlusDataError(boost::algorithm::join(m_impl->m_errors, ", "));
  }
  m_impl->m_enabled = true;
}

void SokobanPlus::disable() {
  m_impl->m_enabled = false;
}

piece_id_t SokobanPlus::box_plus_id(piece_id_t for_id) const {
  try {
    return m_impl->get_plus_id(for_id, m_impl->m_box_plus_ids);
  }
  catch (const out_of_range&) {
    throw KeyError("No box with ID: " + std::to_string(for_id));
  }
}

piece_id_t SokobanPlus::goal_plus_id(piece_id_t for_id) const {
  try {
    return m_impl->get_plus_id(for_id, m_impl->m_goal_plus_ids);
  }
  catch (const out_of_range&) {
    throw KeyError("No goal with ID: " + std::to_string(for_id));
  }
}

} // namespace sokoengine

File: src/sokoenginepy/manager/sokoban_plus.py
from .. import utilities
from .piece import DEFAULT_PIECE_ID


class SokobanPlusDataError(ValueError):
    pass


class SokobanPlus:
    """
    Manages Sokoban+ data for game board.

    **Sokoban+ rules**

    In this variant of game rules, each box and each goal on board get number tag
    (color). Game objective changes slightly: board is considered solved only when
    each goal is filled with box of the same tag. So, for example goal tagged with
    number 9 must be filled with any box tagged with number 9.

    Multiple boxes and goals may share same plus id, but the number of boxes with one
    plus id must be equal to number of goals with that same plus id. There is also
    default plus id that represents non tagged boxes and goals.

    Sokoban+ ids for given board are defined by two strings called goalorder and
    boxorder. For example, boxorder "13 24 3 122 1" would give plus_id = 13 to box id
    = 1, plus_id = 24 to box ID = 2, etc...

    **Valid Sokoban+ id sequences**

    Boxorder and goalorder must define ids for equal number of boxes and goals. This
    means that in case of boxorder assigning plus id "42" to two boxes, goalorder
    must also contain number 42 twice.

    Sokoban+ data parser accepts any positive integer as plus id.

    Attributes:
        DEFAULT_PLUS_ID: Sokoban+ ID for pieces that don't have one or when
            Sokoban+ is disabled.

            Original Sokoban+ implementation used number 99 for default plus ID. As
            there can be more than 99 boxes on board, sokoenginepy changes this
            detail and uses :const:`DEFAULT_PLUS_ID` as default plus ID. When loading
            older puzzles with Sokoban+, legacy default value is converted
            transparently.

    Args:
        boxorder (str): Space separated integers describing Sokoban+ IDs for boxes
        goalorder (str): Space separated integers describing Sokoban+ IDs for goals
        pieces_count (int): Total count of boxes/goals on board
    """

    _LEGACY_DEFAULT_PLUS_ID = 99
    DEFAULT_PLUS_ID = 0

    def __init__(self, pieces_count, boxorder=None, goalorder=None):
        self._is_enabled = False
        self._is_validated = False
        self.errors = []

        self._pieces_count = pieces_count
        self._box_plus_ids = None
        self._goal_plus_ids = None
        self._boxorder = None
        self._goalorder = None

        self.boxorder = boxorder or ""
        self.goalorder = goalorder or ""

    @classmethod
    def is_sokoban_plus_string(cls, line):
        return utilities.contains_only_digits_and_spaces(
            line
        ) and not utilities.is_blank(line)

    @classmethod
    def is_valid_plus_id(cls, plus_id):
        return isinstance(plus_id, int) and plus_id >= cls.DEFAULT_PLUS_ID

    @property
    def pieces_count(self):
        return self._pieces_count

    @pieces_count.setter
    def pieces_count(self, rv):
        if rv != self._pieces_count:
            self.is_enabled = False
            self._is_validated = False
            self._pieces_count = int(rv)

    @property
    def boxorder(self):
        if self.is_enabled and self.is_valid:
            return self._rstrip_default_plus_ids(
                " ".join(str(i) for i in self._box_plus_ids.values())
            )
        else:
            return self._boxorder

    @boxorder.setter
    def boxorder(self, rv):
        if rv != self._boxorder:
            self.is_enabled = False
            self._is_validated = False
            self._boxorder = rv or ""

    @property
    def goalorder(self):
        if self.is_enabled and self.is_valid:
            return self._rstrip_default_plus_ids(
                " ".join(str(i) for i in self._goal_plus_ids.values())
            )
        else:
            return self._goalorder

    @goalorder.setter
    def goalorder(self, rv):
        if rv != self._goalorder:
            self.is_enabled = False
            self._is_validated = False
            self._goalorder = rv or ""

    @property
    def is_valid(self):
        if self._is_validated:
            return not self.errors

        self.errors = []
        try:
            self._box_plus_ids = self._parse_and_clean_ids_string(self._boxorder)
            self._goal_plus_ids = self._parse_and_clean_ids_string(self._goalorder)
        except ValueError as exc:
            self.errors.append(str(exc))

        self._validate_plus_ids(self._box_plus_ids)
        self._validate_plus_ids(self._goal_plus_ids)
        self._validate_piece_count()
        self._validate_ids_counts()
        self._validate_id_sets_equality()
        self._is_validated = True

        return not self.errors

    @property
    def is_enabled(self):
        return self._is_enabled

    @is_enabled.setter
    def is_enabled(self, value):
        """
        Raises:
            :exc:`SokobanPlusDataError`: Trying to enable invalid Sokoban+
        """
        if value:
            if not self.is_valid:
                raise SokobanPlusDataError(self.errors)
        self._is_enabled = value

    def box_plus_id(self, for_box_id):
        """
        Get Sokoban+ ID for box.

        Args:
            for_box_id (int): box ID

        Returns:
            int: If Sokoban+ is enabled returns Sokoban+ ID of a box. If not, returns
            :const:`DEFAULT_PLUS_ID`

        Raises:
            :exc:`KeyError`: No box with ID ``for_box_id``, but only if i Sokoban+ is
                enabled
        """
        try:
            return self._get_plus_id(for_box_id, from_where=self._box_plus_ids)
        except KeyError:
            raise KeyError("No box with ID: {0}".format(for_box_id))

    def goal_plus_id(self, for_goal_id):
        """
        Get Sokoban+ ID for goal.

        Args:
            for_goal_id (int): goal ID

        Returns:
            int: If Sokoban+ is enabled returns Sokoban+ ID of a goal. If not,
            returns :const:`DEFAULT_PLUS_ID`

        Raises:
            :exc:`KeyError`: No goal with ID ``for_goal_id``, but only if Sokoban+ is
                enabled
        """
        try:
            return self._get_plus_id(for_goal_id, from_where=self._goal_plus_ids)
        except KeyError:
            raise KeyError("No goal with ID: {0}".format(for_goal_id))

    def _rstrip_default_plus_ids(self, plus_ids_str):
        # TODO: Might not work correctly for "3 5 4 6 2 19" or "3 5 4 6 2 10"
        if self.pieces_count < self._LEGACY_DEFAULT_PLUS_ID:
            return plus_ids_str.rstrip(
                str(self.DEFAULT_PLUS_ID) + " " + str(self._LEGACY_DEFAULT_PLUS_ID)
            )
        else:
            return plus_ids_str.rstrip(str(self.DEFAULT_PLUS_ID) + " ")

    def _get_plus_id(self, for_id, from_where):
        if not self.is_enabled:
            return self.DEFAULT_PLUS_ID
        else:
            return from_where[for_id]

    def _parse_and_clean_ids_string(self, plus_ids_str):
        """
        Safely replaces legacy default plus ids with default ones.

        Returns:
            dict: dict that maps piece IDs to piece Sokoban+ IDs
        """

        def convert_or_raise(id_str):
            try:
                return int(id_str)
            except ValueError:
                raise SokobanPlusDataError(
                    "Can't parse Sokoban+ string! Illegal characters found. "
                    "Only digits and spaces allowed."
                )

        trimmed = [
            convert_or_raise(id_str)
            for id_str in self._rstrip_default_plus_ids(plus_ids_str).split()
        ]

        cleaned = [
            self.DEFAULT_PLUS_ID
            if (
                i == self._LEGACY_DEFAULT_PLUS_ID
                and self.pieces_count < self._LEGACY_DEFAULT_PLUS_ID
            )
            else i
            for i in trimmed
        ]

        expanded = cleaned + [self.DEFAULT_PLUS_ID] * (self.pieces_count - len(cleaned))

        retv = dict()
        for index, plus_id in enumerate(expanded):
            retv[DEFAULT_PIECE_ID + index] = plus_id

        return retv

    def _validate_plus_ids(self, ids):
        if ids:
            for i in ids.values():
                if not self.is_valid_plus_id(i):
                    self.errors.append("Invalid Sokoban+ ID: {0}".format(i))

    def _validate_piece_count(self):
        if self.pieces_count < 0:
            self.errors.append("Sokoban+ can't be applied to zero pieces count.")

    def _validate_ids_counts(self):
        if self._box_plus_ids and len(self._box_plus_ids) != self.pieces_count:
            self.errors.append(
                "Sokoban+ boxorder data doesn't contain same amount of IDs "
                + "as there are pieces on board! (pieces_count: {0})".format(
                    self.pieces_count
                )
            )

        if self._goal_plus_ids and len(self._goal_plus_ids) != self.pieces_count:
            self.errors.append(
                "Sokoban+ goalorder data doesn't contain same amount of IDs "
                + "as there are pieces on board! (pieces_count: {0})".format(
                    self.pieces_count
                )
            )

    def _validate_id_sets_equality(self):
        if self._box_plus_ids:
            boxes = set(
                pid
                for pid in self._box_plus_ids.values()
                if pid != self.DEFAULT_PLUS_ID
            )
        else:
            boxes = set()

        if self._goal_plus_ids:
            goals = set(
                pid
                for pid in self._goal_plus_ids.values()
                if pid != self.DEFAULT_PLUS_ID
            )
        else:
            goals = set()

        if boxes != goals:
            self.errors.append(
                "Sokoban+ data doesn't define equal sets of IDs for "
                + "boxes and goals"
            )

    def __repr__(self):
        return (
            "SokobanPlus("
            + "pieces_count={0}, boxorder='{1}', goalorder='{2}')".format(
                self.pieces_count, self.boxorder, self.goalorder
            )
        )

    def __str__(self):
        return (
            "SokobanPlus("
            + "pieces_count={0}, boxorder='{1}', goalorder='{2}')".format(
                self.pieces_count, self.boxorder, self.goalorder
            )
        )
